File: src/Beacon.Android/app/src/main/java/dev/beacon/android/AnnexBAccessUnitSplitter.java
package dev.beacon.android;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

final class AnnexBAccessUnitSplitter {
    private static final String MissingStartCode = "Encoded video bytes are not Annex B start-code delimited.";

    private AnnexBAccessUnitSplitter() {
    }

    static List<byte[]> split(byte[] bytes) {
        if (bytes == null || bytes.length < 5 || !startsWithStartCode(bytes, 0)) {
            throw new IllegalArgumentException(MissingStartCode);
        }

        List<Integer> starts = findStartCodes(bytes);
        if (starts.isEmpty()) {
            throw new IllegalArgumentException(MissingStartCode);
        }

        List<byte[]> samples = new ArrayList<>();
        int sampleStart = starts.get(0);
        boolean currentSampleHasVcl = false;

        for (int startIndex = 0; startIndex < starts.size(); startIndex++) {
            int start = starts.get(startIndex);
            int payload = start + 4;
            boolean vcl = payload < bytes.length && isVclNal(bytes[payload]);
            if (vcl && currentSampleHasVcl) {
                samples.add(Arrays.copyOfRange(bytes, sampleStart, start));
                sampleStart = start;
            }

            if (vcl) {
                currentSampleHasVcl = true;
            }
        }

        if (sampleStart < bytes.length) {
            samples.add(Arrays.copyOfRange(bytes, sampleStart, bytes.length));
        }

        return samples;
    }

    private static List<Integer> findStartCodes(byte[] bytes) {
        List<Integer> starts = new ArrayList<>();
        for (int index = 0; index <= bytes.length - 4; index++) {
            if (startsWithStartCode(bytes, index)) {
                starts.add(index);
                index += 3;
            }
        }

        return starts;
    }

    private static boolean startsWithStartCode(byte[] bytes, int index) {
        return bytes[index] == 0 && bytes[index + 1] == 0 && bytes[index + 2] == 0 && bytes[index + 3] == 1;
    }

    private static boolean isVclNal(byte value) {
        int type = value & 0x1F;
        return type >= 1 && type <= 5;
    }
}

File: src/Beacon.Server/Api/AnnexBAccessUnitSplitter.cs
namespace Beacon.Server.Api;

internal static class AnnexBAccessUnitSplitter
{
    private const string MissingStartCode = "Encoded video bytes are not Annex B start-code delimited.";

    public static IReadOnlyList<byte[]> Split(byte[] bytes)
    {
        if (bytes.Length < 4 || GetStartCodeLength(bytes, 0) == 0)
        {
            throw new InvalidOperationException(MissingStartCode);
        }

        List<AnnexBStartCode> starts = FindStartCodes(bytes);
        if (starts.Count == 0)
        {
            throw new InvalidOperationException(MissingStartCode);
        }

        List<byte[]> samples = [];
        int sampleStart = starts[0].Index;
        bool currentSampleHasVcl = false;

        foreach (AnnexBStartCode start in starts)
        {
            int payload = start.Index + start.Length;
            bool vcl = payload < bytes.Length && IsVclNal(bytes[payload]);
            if (vcl && currentSampleHasVcl)
            {
                samples.Add(bytes[sampleStart..start.Index]);
                sampleStart = start.Index;
            }

            currentSampleHasVcl |= vcl;
        }

        if (sampleStart < bytes.Length)
        {
            samples.Add(bytes[sampleStart..]);
        }

        return samples;
    }

    private static List<AnnexBStartCode> FindStartCodes(byte[] bytes)
    {
        var starts = new List<AnnexBStartCode>();
        for (int index = 0; index <= bytes.Length - 3; index++)
        {
            int length = GetStartCodeLength(bytes, index);
            if (length > 0)
            {
                starts.Add(new AnnexBStartCode(index, length));
                index += length - 1;
            }
        }

        return starts;
    }

    private static int GetStartCodeLength(byte[] bytes, int index)
    {
        if (index <= bytes.Length - 4 &&
            bytes[index] == 0 &&
            bytes[index + 1] == 0 &&
            bytes[index + 2] == 0 &&
            bytes[index + 3] == 1)
        {
            return 4;
        }

        if (index <= bytes.Length - 3 &&
            bytes[index] == 0 &&
            bytes[index + 1] == 0 &&
            bytes[index + 2] == 1)
        {
            return 3;
        }

        return 0;
    }

    private static bool IsVclNal(byte value)
    {
        int type = value & 0x1F;
        return type >= 1 && type <= 5;
    }

    private readonly record struct AnnexBStartCode(int Index, int Length);
}
